lib/router.js

```diff
@@ -295,43 +298,28 @@ Router.prototype.prefix = function (prefix) {
 Router.prototype.routes = Router.prototype.middleware = function () {
   var router = this;
 
-  var dispatch = function *dispatch(next) {
-    debug('%s %s', this.method, this.path);
+  var dispatch = function dispatch(ctx, next) {
+    debug('%s %s', ctx.method, ctx.path);
 
-    var path = router.opts.routerPath || this.routerPath || this.path;
-    var matched = router.match(path, this.method);
-    var layer, i, ii;
+    var path = router.opts.routerPath || ctx.routerPath || ctx.path;
+    var matched = router.match(path, ctx.method);
+    var layerChain, layer, i;
 
-    if (this.matched) {
-      this.matched.push.apply(this.matched, matched.path);
+    if (ctx.matched) {
+      ctx.matched.push.apply(ctx.matched, matched.path);
     } else {
-      this.matched = matched.path;
+      ctx.matched = matched.path;
     }
 
-    if (matched.pathAndMethod.length) {
-      i = matched.pathAndMethod.length;
-      while (matched.route && i--) {
-        layer = matched.pathAndMethod[i];
-        ii = layer.stack.length;
-        this.captures = layer.captures(path, this.captures);
-        this.params = layer.params(path, this.captures, this.params);
-        debug('dispatch %s %s', layer.path, layer.regexp);
-
-        while (ii--) {
-          if (layer.stack[ii].constructor.name === 'GeneratorFunction') {
-            next = layer.stack[ii].call(this, next);
-          } else {
-            next = Promise.resolve(layer.stack[ii].call(this, next));
-          }
-        }
-      }
-    }
+    if (!matched.route) return next();
 
-    if (typeof next.next === 'function') {
-      yield *next;
-    } else {
-      yield next;
-    }
+    layerChain = matched.pathAndMethod.reduce(function(memo, layer) {
+      ctx.captures = layer.captures(path, ctx.captures);
+      ctx.params = layer.params(path, ctx.captures, ctx.params);
```

Comment:
Unifying the implementations would be nice. I guess I was wondering if borrowing the 10-15 lines of code and adjusting them to work with layers would work and let kia-router v6 ship since it's one of the biggest blockers of koa2. Could then make changes to koa-compose upstream to accommodate koa-router's use case and then clean up the bit of code here.